### PR TITLE
Implement GetByJobId in gcs table storage

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
## Why are these changes needed?

we wanna remove all related information in table when a job is finished. This could be done by simply invoke methods `deleteByJobId`, but checkpoint table is not related with job id but actor id. So we need implement `getByJobId` interface to get all key/value of actors which is related with job id, then delete all checkpoint data using values got.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/gcs/store_client/in_memory_store_client.cc [hunk 1]
```diff
@@ -89,6 +89,27 @@ Status InMemoryStoreClient::AsyncBatchDelete(const std::string &table_name,
...
+  if (iter != table->index_keys_.end()) {
+    for (auto &key : iter->second) {
+      auto kv_iter = table->records_.find(key);
+      if (kv_iter != table->records_.end()) {
+        result[kv_iter->first] = kv_iter->second;
+      }
+    }
+    main_io_service_.post([result, callback]() { callback(result); });
+  } else {
+    main_io_service_.post([result, callback]() { callback(result); });
```
Comment (raulchen) [inline src/ray/gcs/store_client/in_memory_store_client.cc:108]: can remove `else` and just put this line after the if block

File: src/ray/gcs/gcs_server/gcs_table_storage.cc [hunk 2]
```diff
@@ -85,9 +85,17 @@ Status GcsTableWithJobId<Key, Data>::Put(const Key &key, const Data &value,
...
+  auto on_done = [callback](const std::unordered_map<std::string, std::string> &result) {
+    std::unordered_map<Key, Data> values;
+    for (auto &item : result) {
+      Data data;
+      data.ParseFromString(item.second);
+      values[Key::FromBinary(item.first)] = std::move(data);
+    }
+    callback(values);
+  };
+  return this->store_client_->AsyncGetByIndex(this->table_name_, job_id.Binary(),
```
Comment (raulchen) [inline src/ray/gcs/gcs_server/gcs_table_storage.cc:97]:
```suggestion
  return this->store_client_->AsyncGetByIndex(this->table_name_, job_id.Hex(),
```

File: src/ray/gcs/store_client/in_memory_store_client.cc (hunk 1 above)
Comment (WangTaoTheTonic) [reply src/ray/gcs/store_client/in_memory_store_client.cc]: ok!

File: src/ray/gcs/gcs_server/gcs_table_storage.cc (hunk 2 above)
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_server/gcs_table_storage.cc:97]: the rest methods like Put/DeleteByJobId both use `Binary`, so we need keep them consistent. Anyway we can change them to `Hex` together in future :)

**Review by raulchen — APPROVED** [review]

**Pull request merged**